odl/space/base_ntuples.py

```diff
@@ -403,6 +403,281 @@ def __repr__(self):
+        6.0
+        >>> np.add.reduce(x, keepdims=True)
+        rn(1).element([6.0])
+
+        Finally, ``add.reduceat`` is a combination of ``reduce`` and
+        ``at`` with rather flexible and complex semantics (see the
+        `reduceat documentation`_ for details):
+
+        >>> np.add.reduceat(x, [0, 1])
+        rn(2).element([1.0, 5.0])
+
+        References
+        ----------
+        .. _corresponding NEP:
+           https://github.com/numpy/numpy/blob/master/doc/neps/\
+ufunc-overrides.rst
+
+        .. _interface documentation:
+           https://github.com/charris/numpy/blob/master/doc/source/reference/\
+arrays.classes.rst#special-attributes-and-methods
+
+        .. _general documentation on Numpy ufuncs:
+           https://docs.scipy.org/doc/numpy/reference/ufuncs.html
+
+        .. _reduceat documentation:
+           https://docs.scipy.org/doc/numpy/reference/generated/\
+numpy.ufunc.reduceat.html
+        """
+        # Unwrap out if provided. The output parameters are all wrapped
+        # in one tuple, also if there is only one.
+        out_tuple = kwargs.pop('out', ())
+
+        # We allow our own vectors and `numpy.ndarray` objects as `out`
+        if not all(isinstance(out, (type(self), np.ndarray)) or out is None
+                   for out in out_tuple):
+            return NotImplemented
+
+        # Convert inputs that are ODL vectors to arrays so that the
+        # native Numpy ufunc is called later
+        inputs = tuple(
+            inp.asarray() if isinstance(inp, type(self)) else inp
+            for inp in inputs)
+
+        out = out1 = out2 = None
+        if len(out_tuple) == 1:
+            out = out_tuple[0]
+        elif len(out_tuple) == 2:
+            out1 = out_tuple[0]
+            out2 = out_tuple[1]
+
+        # Use some of the kwargs for `writable_array`
+        # TODO: propagate `order` when tensors are available
+        array_kwargs = {}
+        out_dtype = kwargs.get('dtype', None)
+        if out_dtype is not None:
+            array_kwargs['dtype'] = out_dtype
+
+        # Need new space for ufunc if dtype was changed
+        # TODO: use `shape` and `order` when tensors are available
+        if out_dtype is None or out_dtype == self.dtype:
+            out_space = self.space
+        else:
+            out_space = type(self.space)(self.size, out_dtype)
+
+        if method == '__call__':
+            if ufunc.nout == 1:
+                if out is None:
+                    out = out_space.element()
+                with writable_array(out, **array_kwargs) as out_arr:
+                    kwargs['out'] = (out_arr,)
+                    ufunc(*inputs, **kwargs)
+                return out
+
+            elif ufunc.nout == 2:
+                if out1 is None:
+                    out1 = out_space.element()
+                if out2 is None:
+                    out2 = out_space.element()
+                with writable_array(out1, **array_kwargs) as out1_arr:
```

Comment:
Use a "multiple with" instead

https://stackoverflow.com/questions/893333/multiple-variables-in-python-with-statement

```python
with A() as a, B() as b, C() as c:
```